pkg/bindinfo/binding_cache.go

```diff
@@ -52,26 +55,43 @@ type fuzzyBindingCache struct {
 	fuzzy2SQLDigests map[string][]string // fuzzyDigest --> sqlDigests
 
 	sql2FuzzyDigest map[string]string // sqlDigest --> fuzzyDigest
+
+	// loadBindingFromStorageFunc is used to load binding from storage if cache miss.
+	loadBindingFromStorageFunc func(sqlDigest string) (Bindings, error)
 }
 
-func newFuzzyBindingCache() FuzzyBindingCache {
+func newFuzzyBindingCache(loadBindingFromStorageFunc func(string) (Bindings, error)) FuzzyBindingCache {
 	return &fuzzyBindingCache{
-		BindingCache:     newBindCache(),
-		fuzzy2SQLDigests: make(map[string][]string),
-		sql2FuzzyDigest:  make(map[string]string),
+		BindingCache:               newBindCache(),
+		fuzzy2SQLDigests:           make(map[string][]string),
+		sql2FuzzyDigest:            make(map[string]string),
+		loadBindingFromStorageFunc: loadBindingFromStorageFunc,
 	}
 }
 
 func (fbc *fuzzyBindingCache) FuzzyMatchingBinding(sctx sessionctx.Context, fuzzyDigest string, tableNames []*ast.TableName) (matchedBinding Binding, isMatched bool) {
+	leastWildcards := len(tableNames) + 1
+	enableFuzzyBinding := sctx.GetSessionVars().EnableFuzzyBinding
+	matchedBinding, leastWildcards, isMatched, missingSQLDigest := fbc.getFromMemory(sctx, fuzzyDigest, tableNames, enableFuzzyBinding, leastWildcards)
+	if len(missingSQLDigest) == 0 {
+		return
+	}
+	if fbc.loadBindingFromStorageFunc == nil {
+		return
+	}
+	fbc.loadFromStore(missingSQLDigest) // loadFromStore's SetBinding has a Mutex inside, so it's safe to call it without lock
+	matchedBinding, _, isMatched, _ = fbc.getFromMemory(sctx, fuzzyDigest, tableNames, enableFuzzyBinding, leastWildcards)
+	return
+}
+
+func (fbc *fuzzyBindingCache) getFromMemory(sctx sessionctx.Context, fuzzyDigest string, tableNames []*ast.TableName, enableFuzzyBinding bool, leastWildcards int) (matchedBinding Binding, leastWildcardsResult int, isMatched bool, missingSQLDigest []string) {
```

Comment:
The argument `leastWildcards` seems useless.

Reply:
updated